core/src/main/scala/chisel3/RawModule.scala

```diff
@@ -183,20 +184,31 @@ abstract class RawModule extends BaseModule {
   }
   private[chisel3] val stagedSecretCommands = collection.mutable.ArrayBuffer[Command]()
 
+  private def secretProbeDefines(left: Data, right: Data)(implicit si: SourceInfo): Seq[Command] = {
+    collectLeafMembers(left).zip(collectLeafMembers(right)).map {
```

Comment:
What if `left` and `right` are not the same type? Is that checked before `secretConnection` is called? You may want to consider using one of the collect functions that iterates over two types at the same tame, eg. `collectMembersOverAll` and then you can error on mismatched fields. All that being said, it really feels like this logic should be common with regular connections. I know you're just tweaking what was already there, but why is this logic necessary vs. reusing something in normal connections?

Reply:
> should be common with regular connections.

Do you have pointers to the code that overlaps? I was rolling my own b/c I wasn't able to find logic that does something similar.

I'm also not sure how to use `collectMembersOverAll`... Do you know where I should be adding in the `HasMatchingZipOfChildren` trait?

cc @azidar in case you have any ideas on how to best integrate this chunk of code

Reply:
I believe there is a `DataMirror.collectMembersOverMatches` here that we should be using rather than the raw zip of the members... @azidar could advise?

Reply:
Yeah, I think it should look something like:

```
collectMembersOverMatches(left, right) {
  case (l, r) => (hasProbeTypeModifier(l), hasProbeTypeModifier(r)) match {
          case (true, true)                              => ProbeDefine(si, l.lref, Node(r))
          case (true, false) if l.probeInfo.get.writable => ProbeDefine(si, l.lref, RWProbeExpr(Node(r)))
          case (true, false)                             => ProbeDefine(si, l.lref, ProbeExpr(Node(r)))
          case (false, true)                             => Connect(si, l.lref, ProbeRead(Node(r)))
          case (false, false)                            => Connect(si, l.lref, Node(r))
    }
}
```

I believe `HasMatchingZipOfChildren` should already be defined?

Reply:
Can you take a look at the new pattern using `collectMembersOverMatches`? E.g.:
```scala
    collectMembersOverMatches(probe.asInstanceOf[Data], value.asInstanceOf[Data]) {
      case (p: Aggregate, v: Aggregate) => // do nothing; only operate on leaves
      case (p, v) => // actual use case
```